### ISSUE TITLE
504 Gateway Timeout errors are not properly handled

### ISSUE THREAD

**Issue by FooBarWidget** [contentful/contentful.net#186]
In the past 2 weeks, the Contentful API servers return many 504 Gateway Timeout errors. Such errors do not have a response body (Content-Length: 0). We submitted a support ticket (ticket ID #43303). Tony from platform support confirmed that this was a server-side problem, which should be mostly fixed now.

Tony also recommended that clients should retry upon receiving a 504 Gateway Timeout. However, contentful.net's current implementation does not perform such retries, and also does not allow callers to implement retries themselves.

## Actual behavior upon receiving a 504

Contentful.net throws a JSON parse error because CreateExceptionForFailedRequest() always assumes that the body is valid JSON. See https://github.com/contentful/contentful.net/blob/v4.1.1/Contentful.Core/ContentfulClientBase.cs#L92

Sample stack trace:

~~~
Newtonsoft.Json.JsonReaderException: Error reading JObject from JsonReader. Path '', line 0, position 0.
   at Newtonsoft.Json.Linq.JObject.Load(JsonReader reader, JsonLoadSettings settings)
   at Newtonsoft.Json.Linq.JObject.Parse(String json, JsonLoadSettings settings)
   at Contentful.Core.ContentfulClientBase.CreateExceptionForFailedRequest(HttpResponseMessage res)
   at Contentful.Core.ContentfulClientBase.EnsureSuccessfulResult(HttpResponseMessage response)
   at Contentful.Core.ContentfulClientBase.SendHttpRequest(HttpRequestMessage request, CancellationToken cancellationToken)
   at Contentful.Core.ContentfulClientBase.SendHttpRequest(String url, HttpMethod method, String authToken, CancellationToken cancellationToken, HttpContent content, Nullable`1 version, String contentTypeId, String organisationId, List`1 additionalHeaders)
   at Contentful.Core.ContentfulManagementClient.GetAsync(String url, CancellationToken cancellationToken, Nullable`1 version, List`1 additionalHeaders)
   at Contentful.Core.ContentfulManagementClient.GetEditorInterface(String contentTypeId, String spaceId, CancellationToken cancellationToken)
~~~

## Requested behavior upon receiving a 504

I think Contentful.net should either throw a proper exception (e.g. a hypothetical `Contentful.GatewayTimeoutException`), or it should perform a retry by itself.

**Comment by Roblinde** [comment]
I do not agree that the SDK should retry on 500 errors, this should always be considered a server error that should be handled separately. There's already retry-handling for 429 errors that can be configured.

I do however agree that it shouldn't be throwing a JsonReaderException, that could be handled better. When (if) I get my access back I'll look into it.

Thanks for the contribution!

**Comment by Roblinde** [comment]
Thanks for the excellent pull request. I have now merged this and will create a release very shortly, including a few other fixes as well.

**Comment by Roblinde** [comment]
Fixed in 4.3.0
